forgerock-auth/src/main/java/org/forgerock/android/auth/devicebind/ApplicationPinDeviceAuthenticator.kt

```diff
@@ -0,0 +1,208 @@
+            this.onPinReceived = {
+                onCredentialsReceived(it.toCharArray())
+            }
+            this.onCancelled = {
+                onCredentialsReceived(CharArray(0))
+            }
+            this.show(fragmentActivity.supportFragmentManager, DeviceBindFragment.TAG)
+        }
+    }
+
+    /**
+     * Retrieve the file input stream to store the [KeyStore]
+     * @param context The Application [Context]
+     * @param keyAlias The suggested key alias
+     */
+    open fun getKeystoreFileInputStream(context: Context, keyAlias: String): FileInputStream {
+        return getEncryptedFile(context, keyAlias).openFileInput();
+    }
+
+    /**
+     * Retrieve the file output stream of the [KeyStore]
+     * @param context The Application [Context]
+     * @param keyAlias The suggested key alias
+     */
+    open fun getKeystoreFileOutputStream(context: Context, keyAlias: String): FileOutputStream {
+        return getEncryptedFile(context, keyAlias).openFileOutput();
+    }
+
+    /**
+     * Retrieve the Keystore Type, default to [KeyStore.getDefaultType]
+     */
+    open fun getKeystoreType(): String {
+        return KeyStore.getDefaultType()
+    }
+
+    private fun persist(context: Context,
+                        keyAlias: String,
+                        keyPair: java.security.KeyPair,
+                        pin: CharArray) {
+        val keyStore = KeyStore.getInstance(getKeystoreType())
+        keyStore.load(null);
+        val privateKeyEntry =
+            PrivateKeyEntry(keyPair.private, arrayOf(generateCertificate(keyPair)))
+        keyStore.setEntry(keyAlias, privateKeyEntry, PasswordProtection(pin))
+        getKeystoreFileOutputStream(context, keyAlias).use {
+            it.flush()
+            keyStore.store(it, null)
+        }
+
+    }
+
+    private fun getPrivateKey(context: Context,
+                              keyAlias: String,
+                              pin: CharArray,
+                              statusResult: (DeviceBindingStatus<PrivateKey>) -> Unit) {
+        if (pin.isNotEmpty()) {
+            try {
+                val keystore = KeyStore.getInstance(getKeystoreType())
+                getKeystoreFileInputStream(context, keyAlias).use {
+                    keystore.load(it, null);
+                }
+                val entry =
+                    keystore.getEntry(keyAware.key, PasswordProtection(pin)) as PrivateKeyEntry
+                statusResult(Success(entry.privateKey))
+            } catch (e: FileNotFoundException) {
+                statusResult(UnRegister())
+                return
+            } catch (e: UnrecoverableKeyException) {
+                statusResult(UnAuthorize())
+            }
+        } else {
+            statusResult(Abort())
+        }
+    }
+
+    private fun getEncryptedFile(context: Context, keyAlias: String): EncryptedFile {
+        val masterKeyAlias = MasterKeys.getOrCreate(MasterKeys.AES256_GCM_SPEC)
```

Comment:
getOrCreate may get deprecated soon in next version i think. do we want to use the  new MasterKey.Builder() to build?

Reply:
Not sure when will Android official release the new version, for now stick with the released version instead of alpha version.